be2-scala/src/main/scala/ch/epfl/pop/storage/DbActorNew.scala

```diff
@@ -1,14 +1,307 @@
+
+
+  /* --------------- Functions handling messages DbActor may receive --------------- */
+
+  private def write(channel: Channel, message: Message): DbActorMessage = {
+    // create channel if missing. If already present => createChannel does nothing
+    val _object = message.decodedData match {
+      case Some(data) => data._object
+      case _ => ObjectType.LAO
+    }
+    createChannel(channel, _object)
+
+    val channelData: ChannelData = readChannelData(channel).channelData
+    storage.write(
+      (channel.toString, channelData.addMessage(message.message_id).toJsonString),
+      (s"$channel:${message.message_id}", message.toJsonString)
+    )
+
+    DbActorAck()
+  }
+
+  private def read(channel: Channel, messageId: Hash): DbActorMessage = ???
+
+  private def readChannelData(channel: Channel): DbActorReadChannelDataAck = ???
+
+  private def readLaoData(channel: Channel): DbActorMessage = ???
+
+  private def catchupChannel(channel: Channel): DbActorMessage = ???
+
+  private def writeAndPropagate(channel: Channel, message: Message): DbActorMessage = {
+    val answer: DbActorMessage = write(channel, message)
+    mediatorRef ! PubSubMediator.Propagate(channel, message)
+    answer
+  }
+
+  private def createChannel(channel: Channel, objectType: ObjectType.ObjectType): DbActorAck = {
+    checkChannelExistence(channel) match {
+      case DbActorAck() => DbActorAck() // do nothing if the channel already exists
+      case _ =>
+        Try(storage.write(channel.toString, ChannelData(objectType, List.empty).toJsonString)) match {
+          case Success(_) => DbActorAck()
+          case Failure(ex) => throw DbActorNAckException(ErrorCodes.SERVER_ERROR.id, ex.getMessage)
+        }
+    }
+  }
+
+  private def createChannels(list: List[(Channel, ObjectType.ObjectType)]): DbActorAck = {
+
+    @scala.annotation.tailrec
+    def filterExistingChannels(
+                                list: List[(Channel, ObjectType.ObjectType)],
+                                acc: List[(Channel, ObjectType.ObjectType)]
+                              ): List[(Channel, ObjectType.ObjectType)] = {
+      list match {
+        case Nil => acc
+        case head :: tail => checkChannelExistence(head._1) match {
+          case DbActorAck() => filterExistingChannels(tail, acc) // already exists in db
+          case _ => filterExistingChannels(tail, head :: acc) // does not exist in db
+        }
+      }
+    }
+
+    // removing channels already present in the db from the list
+    val filtered: List[(Channel, ObjectType.ObjectType)] = filterExistingChannels(list, Nil)
+    // creating ChannelData from the filtered input
+    val mapped: List[(String, String)] = filtered.map { case (c, o) => (c.toString, ChannelData(o, List.empty).toJsonString) }
+
+    Try(storage.write(mapped : _*)) match {
+      case Success(_) => DbActorAck()
+      case Failure(ex) => throw DbActorNAckException(ErrorCodes.SERVER_ERROR.id, ex.getMessage)
+    }
```

Comment:
would it make sense to generalize this "write/ack/nack" logic ? It seems to be the end of most functions.